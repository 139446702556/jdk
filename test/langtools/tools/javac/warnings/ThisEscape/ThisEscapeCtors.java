/*
 * @test /nodynamiccopyright/
 * @bug 8194743
 * @compile/ref=ThisEscapeCtors.out -Xlint:this-escape -XDrawDiagnostics ThisEscapeCtors.java
 * @summary Verify pruning of 'this' escape warnings for various constructors
 */

public class ThisEscapeCtors {

    // This constructor should NOT generate a warning because it would be a
    // duplicate of the warning already generated for ThisEscapeCtors(short).
    public ThisEscapeCtors(char x) {
        this((short)x);
    }

    // This constructor should generate a warning because it invokes leaky this()
    // and is accessible to subclasses.
    public ThisEscapeCtors(short x) {
        this();
    }

    // This constructor should generate a warning because it invokes leaky this()
    // and is accessible to subclasses.
    public ThisEscapeCtors(int x) {
        this();
    }

    // This constructor should NOT generate a warning because it is not accessbile
    // to subclasses. However, other constructors do invoke it, and that should cause
    // them to generate an indirect warning.
    private ThisEscapeCtors() {
        this.mightLeak();
    }

    public void mightLeak() {
    }
}
