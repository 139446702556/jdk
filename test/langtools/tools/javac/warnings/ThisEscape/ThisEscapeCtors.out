ThisEscapeCtors.java:32:23: compiler.warn.possible.this.escape
ThisEscapeCtors.java:19:13: compiler.warn.possible.this.escape.location
ThisEscapeCtors.java:32:23: compiler.warn.possible.this.escape
ThisEscapeCtors.java:25:13: compiler.warn.possible.this.escape.location
4 warnings
